Parameter null-checking analyzer and fixer

RikkiGibson opened the pull request:
Related to #36024

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/UseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,193 @@
+                    return;
+                }
+
+                context.RegisterSyntaxNodeAction(context => AnalyzeSyntax(context, argumentNullException, referenceEqualsMethod), SyntaxKind.MethodDeclaration);
```
Do we want to analyze `LocalFunctionStatement`s as well?

RikkiGibson replied:
That seems reasonable to me. I wonder if there are other declaration kinds we care about. I considered indexers, for example, but it feels not as straightforward, since you essentially need to ensure you can replace equivalent null checks in all of the accessors.

RikkiGibson replied:
Ended up adding constructors as well as different kinds of anonymous methods. I think that's probably everything.

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/UseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = context.SemanticModel!;
+            var syntaxTree = semanticModel.SyntaxTree;
```
Just to avoid null suppression.

```suggestion
            var syntaxTree = context.Node.SyntaxTree;
```

Youssef1313 commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,109 @@
+    internal class CSharpUseParameterNullCheckingCodeFixProvider : SyntaxEditorBasedCodeFixProvider
+    {
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
```
This shouldn't be done for new code I believe.

CyrusNajmabadi replied:
correct.

RikkiGibson replied:
Fixed

Youssef1313 commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,109 @@
+        private class MyCodeAction : CustomCodeActions.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument, title)
```
equivalence key is recommended to be non-localized. The recommendation is to go for `nameof(CSharpAnalyzersResources.Use_parameter_null_checking)`

Youssef1313 commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,109 @@
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
```
There is an indexer I think?

```suggestion
            var diagnostic = context.Diagnostics[0];
```

Youssef1313 commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,364 @@
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UseParameterNullChecking
+{
+    public partial class UseParameterNullCheckingTests : AbstractCSharpDiagnosticProviderBasedUserDiagnosticTest
```
This should use the new test framework.

RikkiGibson replied:
Could you please provide an example of usage of the new framework? I actually just copied all the corresponding files from `UseIsNullCheckForCastAndEqualityOperator` as a basis for this.

Youssef1313 replied:
@RikkiGibson https://github.com/dotnet/roslyn/blob/main/src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs

Youssef1313 commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/CodeStyleOptions2.cs:
```diff
@@ -215,6 +215,12 @@ private static PerLanguageOption2<CodeStyleOption2<bool>> CreateQualifyAccessOpt
             "dotnet_style_prefer_is_null_check_over_reference_equality_method",
             $"TextEditor.%LANGUAGE%.Specific.{nameof(PreferIsNullCheckOverReferenceEqualityMethod)}");
 
+        internal static readonly PerLanguageOption2<CodeStyleOption2<bool>> PreferParameterNullChecking = CreateOption(
```
The poor `AutomationObject` gets forgotten every time a new option is getting added :(
I believe this one should be added just like those:

https://github.com/dotnet/roslyn/blob/ab1062214006506b13f19bfb6215d84f7e71cb0f/src/VisualStudio/CSharp/Impl/Options/AutomationObject/AutomationObject.Style.cs#L234-L244

The effect of this is the user preference being persisted when importing/exporting VS settings.

RikkiGibson replied:
thanks!

Youssef1313 commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,109 @@
+                        };
+
+                        // one of the sides of the binary must be a parameter
+                        var parameterInIf = (IParameterSymbol?)model.GetSymbolInfo(unwrapCast(left), cancellationToken).Symbol
+                            ?? (IParameterSymbol)model.GetSymbolInfo(unwrapCast(right), cancellationToken).Symbol!;
```
null suppression worries me.

RikkiGibson replied:
Is unusual for code fixes to make assumptions about what was found by the analyzer?

Youssef1313 replied:
@RikkiGibson I'm not sure, but I thought it is possible that the code changes between the analyzer reporting a diagnostic and the invocation of the <kbd>CTRL</kbd>+<kbd>.</kbd>

RikkiGibson replied:
I would be surprised if the code fix were invoked under conditions where the diagnostic no longer referred to the same thing that was reported by the analyzer. @CyrusNajmabadi @mavasani do you think you could clarify?

CyrusNajmabadi replied:
yes. t his code can make assumptions that everything must be in a correct state as per the analyzer.  *however* a better approach is to have the fixer just include this information in teh 'AdditionalLocations' section of the diagnostic.  Then teh fixer can just pull it out of the diagnostic and not have to compute anything.

RikkiGibson replied:
This code was deleted, so resolving the thread.

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/UseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,193 @@
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = context.SemanticModel!;
```
SemanticModel isn't declared nullable. Why the suppression is needed?

https://github.com/dotnet/roslyn/blob/ab1062214006506b13f19bfb6215d84f7e71cb0f/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L1404

RikkiGibson replied:
Must have gotten brought in and left in over the course of some refactoring. Thanks

RikkiGibson replied:
Evokes #25372 :)

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/UseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,193 @@
+            var semanticModel = context.SemanticModel!;
+            var syntaxTree = semanticModel.SyntaxTree;
+
+            var option = context.Options.GetOption(CodeStyleOptions2.PreferParameterNullChecking, semanticModel.Language, syntaxTree, cancellationToken);
```
`option.Notification.Severity` should be used to report the diagnostic. See https://github.com/dotnet/roslyn/pull/52077#discussion_r616176581 for info.

RikkiGibson replied:
Thanks. I'm making the adjustment, hopefully it's correct.

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/UseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,241 @@
+            {
+                MethodDeclarationSyntax methodDecl => methodDecl.Body,
+                LocalFunctionStatementSyntax localFunctionStatement => localFunctionStatement.Body,
+                _ => null
```
Could be `throw ExceptionUtilities.Unreachable`?

RikkiGibson replied:
Have made the change, and it helped me catch a bug, so thanks.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(context =>
+            {
+                if (((CSharpCompilation)context.Compilation).LanguageVersion < LanguageVersionExtensions.CSharpNext)
```
PROTOTYPE.  so we change this once you release?

RikkiGibson replied:
The compiler-side `LanguageVersionFacts.CSharpNext` references `LanguageVersionExtensions.CSharpNext`, so I feel confident that when we ship 11 we will remember to fix these up.

CyrusNajmabadi replied:
k.  as long as this is s standard pattern that we know to fixup, this is fine :)

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                    return;
+                }
+
+                var argumentNullExceptionConstructor = argumentNullException.InstanceConstructors.FirstOrDefault(m =>
+                    m.DeclaredAccessibility == Accessibility.Public
+                    && m.Parameters.Length == 1
+                    && m.Parameters[0].Type.SpecialType == SpecialType.System_String);
```
```suggestion
                var argumentNullExceptionConstructor = argumentNullException.InstanceConstructors.FirstOrDefault(m =>
                    m is { DeclaredAccessibility: Accessibility.Public, Parameters.Length: 1 }
                    && m.Parameters[0].Type.SpecialType == SpecialType.System_String);
```

RikkiGibson replied:
Thank you for the suggestion. I prefer the current factoring so will leave as is.

CyrusNajmabadi replied:
sgtm.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                var objectType = compilation.GetSpecialType(SpecialType.System_Object);
+                var referenceEqualsMethod = objectType
+                    .GetMembers(nameof(ReferenceEquals))
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(m => m.DeclaredAccessibility == Accessibility.Public && m.Parameters.Length == 2);
```
```suggestion
                    .FirstOrDefault(m => is IMethodSymbol { DeclaredAccessibility: Accessibility.Public Parameters.Length: 2 });
```

RikkiGibson replied:
Have made the change mainly to reduce number of intermediate iterators.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+            var block = node switch
+            {
+                MethodDeclarationSyntax methodDecl => methodDecl.Body,
+                LocalFunctionStatementSyntax localFunctionStatement => localFunctionStatement.Body,
```
what about lambdas? i guess chance of them doing arg checking is minimal?

RikkiGibson replied:
Have made the change to handle lambdas.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                    .GetMembers(nameof(ReferenceEquals))
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(m => m.DeclaredAccessibility == Accessibility.Public && m.Parameters.Length == 2);
+                // This seems sufficiently rare that it's not a big deal to just not perform analysis in this case.
```
true.... but i also wouldn't mind just contuining in this case and just not looking for RefEquals.  nbd though.

RikkiGibson replied:
have made the change to tolerate missing ReferenceEquals, but still need to add a test.

RikkiGibson replied:
I can't figure out how to add a test where ReferenceEquals is not defined. Any guidance/suggestions would be appreciated.

Youssef1313 replied:
@RikkiGibson I think this could be done via a "fake" ReferenceAssemblies, like this?

https://github.com/dotnet/roslyn/blob/7f889ac6595990547fc0eb7057ad96d171160512/src/Analyzers/CSharp/Tests/UseIndexOrRangeOperator/UseRangeOperatorTests.cs#L55-L76

The implementation of the called constructor just for reference:

https://github.com/dotnet/roslyn-sdk/blob/9e0d174c165bccdc18f8231c4268b1a19f2be420/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs#L43-L52

sharwell replied:
I don't think there's any value in special-casing the analyzer for this situation.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                context.ReportDiagnostic(DiagnosticHelper.Create(Descriptor, statement.GetLocation(), option.Notification.Severity, additionalLocations: null, properties: null));
+            }
+
+            bool areOperandsApplicable(ExpressionSyntax maybeParameter, ExpressionSyntax maybeNullLiteral, [NotNullWhen(true)] out IParameterSymbol? parameter)
```
```suggestion
            return;

            bool AreOperandsApplicable(ExpressionSyntax maybeParameter, ExpressionSyntax maybeNullLiteral, [NotNullWhen(true)] out IParameterSymbol? parameter)
```

CyrusNajmabadi replied:
we use lowercase to distinguish an instance of a delegate from a method/local-func.  this way we can see by name alone if there are allocs involved.

RikkiGibson replied:
That's interesting. Thanks. Have made the change.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                return isParameter(maybeParameter, out parameter);
+            }
+
+            bool isParameter(ExpressionSyntax maybeParameter, [NotNullWhen(true)] out IParameterSymbol? parameter)
```
```suggestion
            bool IsParameter(ExpressionSyntax maybeParameter, [NotNullWhen(true)] out IParameterSymbol? parameter)
```

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                return true;
+            }
+
+            bool isConstructorApplicable(ObjectCreationExpressionSyntax exceptionCreation, IParameterSymbol parameterSymbol)
```
```suggestion
            bool IsConstructorApplicable(ObjectCreationExpressionSyntax exceptionCreation, IParameterSymbol parameterSymbol)
```

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                                right = patternExpression;
+                                break;
+                            case { Condition: InvocationExpressionSyntax { Expression: var receiver, ArgumentList.Arguments: { Count: 2 } arguments } }:
+                                if (!referenceEqualsMethod.Equals(semanticModel.GetSymbolInfo(receiver, cancellationToken).Symbol))
```
consider a when-clause.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                                continue;
+                        }
+
+                        IParameterSymbol? parameter;
+                        if (!areOperandsApplicable(left, right, out parameter)
+                            && !areOperandsApplicable(right, left, out parameter))
```
```suggestion
                        if (!AreOperandsApplicable(left, right, out var parameter)
                            && !AreOperandsApplicable(right, left, out parameter))
```

though this comes down to taste.  leave if you prefer as is (for the parameter decl).

RikkiGibson replied:
Code has changed for unrelated reasons, so resolving the thread.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                            _ => null
+                        };
+
+                        if (throwStatement is null
+                            || throwStatement.Expression is not ObjectCreationExpressionSyntax thrownInIf
+                            || !isConstructorApplicable(thrownInIf, parameter))
```
```suggestion
                        if (throwStatement?.Expression is not ObjectCreationExpressionSyntax thrownInIf
                            || !isConstructorApplicable(thrownInIf, parameter))
```

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                        continue;
+                }
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(Descriptor, statement.GetLocation(), option.Notification.Severity, additionalLocations: null, properties: null));
```
i'm surpirsed the statement location is good enough.  i would think you'd want the statement location and the parameter location (in additional locations).  that way you won't have to re-infer the parameter in question when fixing up the statemnet.

RikkiGibson replied:
I did have to re-infer the parameter that should be null checked. I'm happy to make the implementation use AdditionalLocations if you feel like that's the better way.

CyrusNajmabadi replied:
yup.  it's the standard pattern and just makes things easier :)

RikkiGibson replied:
Have made the change.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+            {
+                if (maybeParameter is CastExpressionSyntax { Type: var type, Expression: var operand })
+                {
+                    if (semanticModel.GetTypeInfo(type).Type?.SpecialType != SpecialType.System_Object)
```
doc this for future readers.

CyrusNajmabadi replied:
resolved

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+            {
+                switch (statement)
+                {
+                    // if (param == null) { throw new ArgumentNullException(nameof(param)); }
```
hrmm... i think you need to go further.  except for strings, i think you need to check that this is not an overloaded operator right?  because `!!` wont' be legal.  def test cases liek that, as well as interesting cases of ref-types, value-types, nullable-types, typeparameter-types (constrained and otherwise).

RikkiGibson replied:
As a heuristic, I would consider saying that if the `==` is defined on a special type, then treat it as equivalent to the null check we perform via `!!`. For example: string is a special type, so we don't mind that a non-`object ==` operator is used.

CyrusNajmabadi replied:
does this work for string though?  we likely would want this feature to light up, and strings don't have a built-in (unless i'm mistaken).

RikkiGibson replied:
It does actually do the thing we want on string (string.operator == appears to be considered built in)

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,242 @@
+                {
+                    return false;
+                }
+                var constantValue = semanticModel.GetConstantValue(argument.Expression, cancellationToken);
```
```suggestion

                var constantValue = semanticModel.GetConstantValue(argument.Expression, cancellationToken);
```

RikkiGibson replied:
Have fixed the formatting.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,507 @@
+        private static readonly CSharpParseOptions s_parseOptions = CSharpParseOptions.Default.WithLanguageVersion(LanguageVersionExtensions.CSharpNext);
+
+        public UseParameterNullCheckingTests(ITestOutputHelper logger)
+          : base(logger)
```
```suggestion
            : base(logger)
```

(or on same line as previous).  surprised this didn't fail linting

RikkiGibson replied:
deleting this as part of moving to the new test framework.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,507 @@
+        [Theory]
+        [Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        [InlineData("==")]
+        [InlineData("is")]
```
i'll allow this. :)

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,507 @@
+    private readonly string s;
+    void M(string s)
+    {
+        if ([||]object.ReferenceEquals(s, null))
```
add test with `C.ReferenceEquals(s, null)`  (also are tehre reverse forms of all of these?  e.g. `ReferenceEquals(null, s)`)?

RikkiGibson replied:
I think the current level of testing for the "reverse" form is adequate, but less thorough than the "regular" form with null on the right. I did add a "null on the left" case for ReferenceEquals in response to this feedback, though.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,507 @@
+    void M(string HelloWorld)
+    {
+        if ([||]HelloWorld == null)
+            throw new ArgumentNullException(""Hello"" + ""World"");
```
lol nice.

CyrusNajmabadi commented on src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:
```diff
@@ -166,6 +166,7 @@ internal static class IDEDiagnosticIds
         public const string SimplifyPropertyPatternDiagnosticId = "IDE0170";
 
         public const string UseTupleSwapDiagnosticId = "IDE0180";
+        public const string UseParameterNullCheckingId = "IDE0181";
```
190.  we are trying to give some room to items so taht if we add more cases for them their IDs can be grouped nicely.  that's why we have the jumps from 100/110/120/130/etc.  thanks!

RikkiGibson replied:
Have made the change.

Youssef1313 replied:
@CyrusNajmabadi I'm a bit confused. It seems like we're always making jumps regardless and are missing the purpose of jumps in the first place. To be specific, why not group this new diagnostic with `UseNullCheckOverTypeCheckDiagnosticId`? Anything I'm missing about the idea of "grouping"?

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/CodeStyleOptions2.cs:
```diff
@@ -215,6 +215,12 @@ private static PerLanguageOption2<CodeStyleOption2<bool>> CreateQualifyAccessOpt
+            CodeStyleOptionGroups.ExpressionLevelPreferences, nameof(PreferParameterNullChecking),
+            defaultValue: TrueWithSuggestionEnforcement,
+            "dotnet_style_prefer_parameter_null_checking",
+            $"TextEditor.%LANGUAGE%.Specific.{nameof(PreferParameterNullChecking)}");
```
probably not done yet.  but we need to update the VS options UI to show this.  @jmarolf do we need to updat ethe editorconfig UI as well?

Youssef1313 replied:
@CyrusNajmabadi I've been thinking for veeery loong to have a source generator help in this area. Do you think this is something interesting? I may try to work on it in free time.

CyrusNajmabadi replied:
Works for me

RikkiGibson replied:
I am going through the boilerplate now and yeahhh I would love for a source generator to make all this go away.

RikkiGibson commented:
Pushing my work for the afternoon. I think there's still more comments to address, expect to wrap it up tomorrow.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+                ConstructorDeclarationSyntax constructorDecl => constructorDecl.Body,
+                LocalFunctionStatementSyntax localFunctionStatement => localFunctionStatement.Body,
+                AnonymousFunctionExpressionSyntax anonymousFunction => anonymousFunction.Block,
+                _ => throw ExceptionUtilities.UnexpectedValue(node)
```
another test case (cuz i'm evil):

```c#
public C(string s)
    => _s = s ?? throw new ArgumentNullException();
```

RikkiGibson replied:
I'd like to support this eventually, as well as `param ?? throw ...` in the `base()`/`this()` constructor initializer, but I'd like to punt on it for this PR so I can switch my focus back to the compiler side issues with the feature temporarily. Does that work for you?

CyrusNajmabadi replied:
it does.  just leave note in the PR.  the important distinction is that this a known thing we're punting on, as opposed to an unintentional omission.  without a comment, there is no way to tell.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+                    Descriptor,
+                    statement.GetLocation(),
+                    option.Notification.Severity,
+                    additionalLocations: new[] { parameter.Locations[0] },
```
for paranoia's sake.  prior to reporting the diag.  ensure:

1. the param has a location.
2. the location's *tree* is the same as our current tree

If so, then you can report this.  

this is likely not necessary, but possibly good in case some error condition makes the compiler collapse symbols in different files.

Youssef1313 replied:
@RikkiGibson Before taking the suggestion, can you add a test similar to this?

```csharp
class C
{
    private int _p;

    public int P
    {
        get => _p;
        set
        {
            local();

            void local()
            {
                _p = value;
            }
        }
    }
}
```

RikkiGibson replied:
I'm adding the suggested test. It passes without issuing any diagnostics, as it should. I do think it's still fine to add the additional validation that Cyrus suggested here.

Youssef1313 replied:
Yup the extra validation doesn't harm and is more future-proof.

RikkiGibson replied:
> 2\. the location's _tree_ is the same as our current tree

I'm assuming this means that the statement should be in the same syntax tree as the parameter.

CyrusNajmabadi replied:
Yup

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+
+            bool IsConstructorApplicable(ObjectCreationExpressionSyntax exceptionCreation, IParameterSymbol parameterSymbol)
+            {
+                if (exceptionCreation.ArgumentList is null
+                    || exceptionCreation.ArgumentList.Arguments.FirstOrDefault() is not { } argument)
```
```suggestion
                if (exceptionCreation.ArgumentList?.Arguments.FirstOrDefault() is not { } argument)
```

?

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+                }
+
+                var constantValue = semanticModel.GetConstantValue(argument.Expression, cancellationToken);
+                if (!constantValue.HasValue || constantValue.Value is not string constantString || !string.Equals(constantString, parameterSymbol.Name, StringComparison.Ordinal))
```
```suggestion
                if (constantValue.Value is not string constantString || !string.Equals(constantString, parameterSymbol.Name, StringComparison.Ordinal))
```

CyrusNajmabadi commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,84 @@
+                switch (node)
+                {
+                    case IfStatementSyntax ifStatement:
+                        editor.RemoveNode(ifStatement);
```
test of removal with a pp directive before/after the removed statement.

CyrusNajmabadi commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,84 @@
+                        throw ExceptionUtilities.UnexpectedValue(node);
+                }
+
+                var parameterSyntax = (ParameterSyntax)diagnostic.AdditionalLocations[0].FindNode(cancellationToken);
```
if you're going to presuem this is a ParameterSyntax, also do a check that it's that type in the analyzer.  note: i might recommend using IParameterSymbol.DeclaringSyntaxReferences (instead of Locations).  The former tends to correspond directly to the syntax node (where the latter often refers to token names and can be iffy on recovery).

RikkiGibson replied:
have added the check

https://github.com/dotnet/roslyn/blob/bdda177645ef4df04ab8044a53738708ac771cd4/src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs#L187-L197

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterCompilationStartAction(context =>
+            {
+                if (((CSharpCompilation)context.Compilation).LanguageVersion < LanguageVersionExtensions.CSharpNext)
+                {
+                    return;
+                }
+
+                var compilation = context.Compilation;
```
A very minor simplification

```suggestion
                var compilation = (CSharpCompilation)context.Compilation;
                if (compilation.LanguageVersion < LanguageVersionExtensions.CSharpNext)
                {
                    return;
                }

```

CyrusNajmabadi replied:
I feel like we might have an extension for this somewhere...  If not, we should just make one :-)

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+            var methodSymbol = node is AnonymousFunctionExpressionSyntax
+                ? semanticModel.GetSymbolInfo(node, cancellationToken).Symbol
+                : semanticModel.GetDeclaredSymbol(node, cancellationToken);
+            if (methodSymbol is null)
```
No functional change intended, but just a quick return instead of unnecessarily analyzing the Block statements.

```suggestion
            if (methodSymbol is null || methodSymbol.Parameters.Length == 0)
```

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+namespace Microsoft.CodeAnalysis.CSharp.UseParameterNullChecking
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpUseParameterNullCheckingDiagnosticAnalyzer
```
Very very minor 😄 

```suggestion
    internal sealed class CSharpUseParameterNullCheckingDiagnosticAnalyzer
```

Youssef1313 commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -0,0 +1,84 @@
+namespace Microsoft.CodeAnalysis.CSharp.UseParameterNullChecking
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseIsNullCheck), Shared]
+    internal class CSharpUseParameterNullCheckingCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```
Minor

```suggestion
    internal sealed class CSharpUseParameterNullCheckingCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```

Youssef1313 commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,728 @@
+    private readonly string s;
+    void M(string s)
+    {
+        [|if (object.ReferenceEquals(s, null))
+        {
+            throw new ArgumentNullException(nameof(s));
+        }|]
```
The diagnostic span is too large that I think it could be noisy. I think reporting on the ParameterSyntax is less nosiy, and can actually much simplify the codefix implementation. @CyrusNajmabadi What do you think?

RikkiGibson replied:
Do you mean reporting on the parameter declaration or the usage?

Youssef1313 replied:
I was thinking of the declaration (ie, swap the diagnostic location with the `additionalLocations`). Implementation-wise, there shouldn't be much of changes. The benefit is mainly a smaller span which I *think* is better user experience. But I'm not sure if there are any guidelines the IDE team uses around diagnostic spans.

CyrusNajmabadi replied:
We discussed it offline. General idea here is that while this is large, it's actually fixed size (not unbounded).

RikkiGibson replied:
I'm happy with it as written because it covers a bounded part of the code that will be eliminated by the fix. Going by that, though, it might be reasonable to change the `field = param ?? throw ...` scenario to squiggle just the `throw ...` part.

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,253 @@
+                        continue;
+                }
+
+                context.ReportDiagnostic(DiagnosticHelper.Create(
+                    Descriptor,
+                    statement.GetLocation(),
+                    option.Notification.Severity,
+                    additionalLocations: new[] { parameter.Locations[0] },
```
Can a parameter have multiple locations? *For example*, for partial methods, does the compiler produce one or two `ParameterSymbol`s?

RikkiGibson replied:
IIRC, A partial method implementation is its own symbol and has its own distinct parameter symbols. It's not a case where the parameter symbol is shared between the two parts and has multiple locations, for example. I'll make sure we have the test, though.

Youssef1313 reviewed (changes requested):
Actually there could be a *possible* bug that can also fix the case indicated in https://github.com/dotnet/roslyn/pull/58182/files#r778339282

```csharp
    public static void M(string s)
    {
        Console.WriteLine(0);
        local();
        void local()
        {
            if (s == null)
                throw new ArgumentNullException(nameof(s)); // Before reporting a diagnostic, we should make sure the parameter syntactically belongs to the same method/local function/etc we're analyzing.
        }
    }
```

Also note that those test cases will need to be tweaked if https://github.com/dotnet/roslyn/pull/58182#discussion_r778157102 is going to be accepted.

RikkiGibson commented:
> Actually there could be a _possible_ bug that can also fix the case indicated in https://github.com/dotnet/roslyn/pull/58182/files#r778339282
> 
> ```cs
>     public static void M(string s)
>     {
>         Console.WriteLine(0);
>         local();
>         void local()
>         {
>             if (s == null)
>                 throw new ArgumentNullException(nameof(s)); // Before reporting a diagnostic, we should make sure the parameter syntactically belongs to the same method/local function/etc we're analyzing.
>         }
>     }
> ```
> 
> Also note that those test cases will need to be tweaked if [#58182 (comment)](https://github.com/dotnet/roslyn/pull/58182#discussion_r778157102) is going to be accepted.

I don't anticipate this being a bug because we check if the parameter's containing symbol is the method being analyzed. However, I'll make sure we have the test.

https://github.com/dotnet/roslyn/blob/b3ce666318947f761acf827f7686236b997ea558/src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs#L220-L224

Youssef1313 commented:
@RikkiGibson Yup I missed this line. A test would be good for coverage anyway :)

RikkiGibson commented:
Have addressed most feedback except for:

- https://github.com/dotnet/roslyn/pull/58182#discussion_r777781255
- https://github.com/dotnet/roslyn/pull/58182#discussion_r778325321

For everything I reacted to with 👍 I'd appreciate if you could check if you consider the comment to be resolved, and if so, either hit "resolve" in the UI or add your own reaction emoji to the comment.

CyrusNajmabadi commented:
I'll review first thing tomorrow

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,260 @@
+
+                context.RegisterSyntaxNodeAction(
+                    context => AnalyzeSyntax(context, argumentNullExceptionConstructor, referenceEqualsMethod),
+                    SyntaxKind.ConstructorDeclaration,
```
Fixing constructors can change the semantics when `base` constructor is called, consider:

```csharp
using System;

new Derived(null);

public class Base
{
    public Base(string s) => Console.WriteLine($"Hi: {s}");
}

public class Derived : Base
{
    public Derived(string s) : base(s)
    {
        if (s == null)
        {
            throw new Exception();
        }
    }
}
```

If a diagnostic is still worth it, then a warning annotation could be there to help. Example:

https://github.com/dotnet/roslyn/blob/7f889ac6595990547fc0eb7057ad96d171160512/src/Features/Core/Portable/ConvertForEachToFor/AbstractConvertForEachToForCodeRefactoringProvider.cs#L56-L57

CyrusNajmabadi replied:
A mutating base, with a derived that threw would be a big anti pattern. I'm ok not caring if semantics change here

Youssef1313 reviewed (commented):
```csharp
using System;
class C
{
    public C(string s) { }
};

class C2 : C
{
    public C2(string s) : base(s ?? throw new ArgumentNullException(nameof(s)))
    {
    }
}
```

A very edge case. Feel free not to handle.

RikkiGibson commented:
> A very edge case. Feel free not to handle.

See https://github.com/dotnet/roslyn/pull/58182#discussion_r778985983

jmarolf commented on src/VisualStudio/Core/Def/EditorConfigSettings/DataProvider/CodeStyle/CommonCodeStyleSettingsProvider.cs:
```diff
@@ -109,6 +109,10 @@ private IEnumerable<CodeStyleSetting> GetNullCheckingCodeStyleOptions(AnalyzerCo
                 description: EditorFeaturesResources.Prefer_is_null_for_reference_equality_checks,
                 editorConfigOptions: options,
                 visualStudioOptions: visualStudioOptions, updater: updater, fileName: FileName);
+            yield return CodeStyleSetting.Create(option: CodeStyleOptions2.PreferParameterNullChecking,
```
👍

jmarolf replied:
I think (since this is a C# only option) this should come from here: https://github.com/dotnet/roslyn/blob/main/src/VisualStudio/CSharp/Impl/EditorConfigSettings/DataProvider/CodeStyle/CSharpCodeStyleSettingsProvider.cs

RikkiGibson replied:
I can't quite figure out the right set of config to use here. Have tried comparing with other csharp specific style options and am scratching my head. Could use some help with someone familiar with this area when possible.

RikkiGibson replied:
Seeing the following test failure with my current config

```log
[xUnit.net 00:00:31.11]     Microsoft.CodeAnalysis.Editor.UnitTests.Diagnostics.ConfigureSeverityLevel.IDEDiagnosticIDConfigurationTests.CSharp_VerifyIDECodeStyleOptionsAreConfigurable [FAIL]
  Failed Microsoft.CodeAnalysis.Editor.UnitTests.Diagnostics.ConfigureSeverityLevel.IDEDiagnosticIDConfigurationTests.CSharp_VerifyIDECodeStyleOptionsAreConfigurable [57 ms]
  Error Message:
   Missing entry:

# IDE0190
No editorconfig based code style option

Expected: False
Actual:   True
  Stack Trace:
     at Microsoft.CodeAnalysis.Editor.UnitTests.Diagnostics.ConfigureSeverityLevel.IDEDiagnosticIDConfigurationTests.<VerifyConfigureCodeStyleOptionsCore>g__ProcessDiagnosticIdAndOption|6_0(String diagnosticId, IOption optionOpt, String editorConfigString, <>c__DisplayClass6_0& ) in /_/src/EditorFeatures/Test/Diagnostics/IDEDiagnosticIDConfigurationTests.cs:line 689
   at Microsoft.CodeAnalysis.Editor.UnitTests.Diagnostics.ConfigureSeverityLevel.IDEDiagnosticIDConfigurationTests.VerifyConfigureCodeStyleOptionsCore(String expected, String languageName) in /_/src/EditorFeatures/Test/Diagnostics/IDEDiagnosticIDConfigurationTests.cs:line 644
   at Microsoft.CodeAnalysis.Editor.UnitTests.Diagnostics.ConfigureSeverityLevel.IDEDiagnosticIDConfigurationTests.CSharp_VerifyIDECodeStyleOptionsAreConfigurable() in /_/src/EditorFeatures/Test/Diagnostics/IDEDiagnosticIDConfigurationTests.cs:line 1047
```

I do not know what I should change here. It seems like `No editorconfig based code style option` is not what we should see in the editorconfig file.

Youssef1313 replied:
https://github.com/dotnet/roslyn/pull/58182#discussion_r780564873 is likely to be the reason :)

jmarolf reviewed (changes requested):
Since this is a C# only option I would make sure that it is only given via the C# style provider. Then everything will light up correctly.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,260 @@
+                    SyntaxKind.LocalFunctionStatement,
+                    SyntaxKind.SimpleLambdaExpression,
+                    SyntaxKind.ParenthesizedLambdaExpression,
+                    SyntaxKind.AnonymousMethodExpression);
```
i think constructor/methods are likely fine.  but you could also extend to anything else that takes parameters (indexers/operators/etc.).  If you dont' want to handle that, please put a note explicitly stating that they're in a "don't care" category, as opposed to an unintentional omission.

RikkiGibson replied:
I'll add a note about indexers. I don't want to support it because of the complexity of potentially eliminating statements from both accessors. Operators should be supported, I'll add a test. I'll make a pass over the method kinds and make sure things are covered.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,260 @@
+
+                    default:
+                        continue;
+                }
```
i recommend extracting this out to a helper that just returns the parameter.  the nested switch/break/continue is pretty hard to erady (for me at least).  tn you.

RikkiGibson replied:
Yes, this leans pretty hard on definite assignment to ensure correctness. I'm open to making the change you've described.

CyrusNajmabadi replied:
i think you could still depend on it.  you'd just have two separate methods for the If vs Assign cases.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+    {
+    }
+}",
+                LanguageVersion = LanguageVersionExtensions.CSharpNext
```
it's too late now (and would be a lot to updatE).  but in teh future, i would have just made a simple test helper that takes in teh before/after.  it would then pass this in, so you'd only have to do this in one location.

RikkiGibson replied:
Makes sense!

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+        [Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        [InlineData("==")]
+        [InlineData("is")]
+        public async Task TestNoBraces(string @operator)
```
do you have a test that verifies you do *not* get the feature prior to c#next?  i would make that the first test btw.

RikkiGibson replied:
See `TestMissingPreCSharp11`

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+        }
+    }
+
+    bool ReferenceEquals(object o1, object o2) => false;
```
interesting case :)

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+                TestCode = testCode,
+                FixedCode = testCode,
+                LanguageVersion = LanguageVersionExtensions.CSharpNext
+            }.RunAsync();
```
i suppose this is ok.  but it is somewhat odd since this would be totally safe to change to `!!`.  but it does seem rare.  we can wait to hear feedback.

RikkiGibson replied:
This imitates the behavior of "use 'is null' instead of '== null'".

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestConstrainedTypeParameter()
```
can you test with constrained on struct?

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+    void M(string s)
+    {
+        if ((object)s == default)
+            throw new ArgumentNullException(nameof(s));
```
interesting.  i think this is ok for now.

RikkiGibson replied:
This was brought in inadvertently because I based my work on the "use 'is null' instead of '== null'" feature. I would be open to making any null constant do the right thing here in a follow up.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestNotWithExceptionNoArguments()
+        {
```
:(   but i'm ok waiting for feedback.

RikkiGibson replied:
Ah. I forgot about this. I can make this happen.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+    static void M(C c)
+    {
+        [|if ((object)c == null)
+            throw new ArgumentNullException(nameof(c));|]
```
can you make this test with 'is' as well (a theory is fine).

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+{
+    static void M(C c!!)
+    {
+#nullable disable
```
so this is bad :)

the point of the pp directive is to get us *not* to remove the leading pp directive.  anotehr good case of this would be `#if blah ... #endif`.  if we remove the former we'll end up with broken code.

RikkiGibson replied:
Does an established pattern exist for preserving structured trivia when we're deleting a node?

CyrusNajmabadi replied:
actually, if you are doing SyntaxEditor.RemoveNode, then that should do the right thing.  can you just add the test that has teh `#if`.  i think it should do the right thing, but the test will verify :)

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,981 @@
+                },
+                LanguageVersion = LanguageVersionExtensions.CSharpNext
+            }.RunAsync();
+        }
```
can you add tests for all other types the compiler supports putting `!!` on (and all other types the compiler would *error* on putting !! on).  For example, if !! is allowed on pointers, can we add a test for that?

i may have missed it, but a test of just a simple value type would be good.  as well as a tests with a nullable value type.

RikkiGibson replied:
I'll make another test pass with this in mind.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/Options/Formatting/StyleViewModel.cs:
```diff
@@ -1122,6 +1122,28 @@ void M2(string value1, string value2)
+//]
+//[
+    // {ServicesVSResources.Over_colon}
+    void M2(string value)
```
```suggestion
    void M1(string value)
```

RikkiGibson replied:
It's not clear to me that this is the right change to make, since other samples in this file use the `M1`/`M2` naming convention.

jmarolf replied:
The intent is that the other section is what happans _after_ this option is applied so as the user flips the option back and forth they can see the difference

CyrusNajmabadi replied:
yes.  this needs to change to M1

RikkiGibson commented:
> Since this is a C# only option I would make sure that it is only given via the C# style provider. Then everything will light up correctly.

I don't fully understand the object model and the differences for language-agnostic versus language-specific style options, but I'll attempt to revert the appropriate added configs to make this happen.

RikkiGibson commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,1416 @@
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestRedundantNullChecks()
+        {
+            // TODO: what test framework settings do we use here?
```
@sharwell I am not certain how to configure the test framework to handle this scenario. When I run this test in its current state I get the following result:

```log
Context: Iterative code fix application
content of '/0/Test0.cs' did not match. Diff shown with expected as baseline:
 
 using System;
 
 class C
 {
     public C(string s!!)
     {
-        if (s is null)
-            throw new ArgumentNullException(nameof(s));
     }
 }

Expected: True
Actual:   False
```

Basically, I would like to just demonstrate that the analyzer/fixer does not do something bad in this scenario, such as crash or produce erroneous code. I don't think it's important to eliminate both marked if-statements in a single iteration of the fixer, for example.

sharwell replied:
The test runs iterations until there is nothing left to fix. It will fail if _any_ iteration causes a problem.

RikkiGibson replied:
Thanks again for the guidance. I think the test is now fixed.

Youssef1313 commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,1416 @@
+{
+    public C(string s!!)
+    {
+        if (s is null)
+            throw new ArgumentNullException(nameof(s));
```
Would the test pass this way?

```suggestion
```

RikkiGibson commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,1410 @@
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestRedundantNullChecks()
+        {
+            // Produces the following error:
```
I updated the test based on @sharwell's advice and now see the error outlined in the comment. I think this is due to replacing the same node multiple times:

https://github.com/dotnet/roslyn/blob/31acd9c7639362ec24dc59e0364da554bb70a5b4/src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs#L69-L70

Could there be some method of avoiding this redundant replacement, while still allowing the parameter to be replaced as expected if any individual instance of the diagnostic is fixed by the user?

RikkiGibson replied:
https://github.com/dotnet/roslyn/pull/58182#discussion_r780502722

Youssef1313 commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -452,7 +452,14 @@ public C(string s!!)
         public async Task TestRedundantNullChecks()
         {
-            // TODO: what test framework settings do we use here?
+            // Produces the following error:
+            //             Message: 
+            // System.InvalidOperationException : GetCurrentNode returned null with the following node: string s - line 321
+            //  Stack Trace: 
+            // Contract.Fail(String message, Int32 lineNumber) line 115
+            // ReplaceChange.Apply(SyntaxNode root, SyntaxGenerator generator) line 321
+            // SyntaxEditor.GetChangedRoot() line 92
+            // <FixAllWithEditorAsync>d__5.MoveNext() line 71
```
https://github.com/dotnet/roslyn/issues/41526 https://github.com/dotnet/roslyn/issues/53960 👀

Youssef1313 replied:
I think the other issues may be a bit different, and I think I got what's happening.

For this scenario, we produced two diagnostics, with the same parameter. The fixer removed the first node and replaced the ParameterSyntax node. Then, it attempts to remove the second node and... when it tries to replace the ParameterSyntax node it doesn't find it. It was already replaced.

```
                var parameterSyntax = (ParameterSyntax)diagnostic.AdditionalLocations[0].FindNode(cancellationToken);
```

We need a way to only try to get this `parameterSyntax` if it wasn't already replaced. Or, consolidate diagnostics related to the same parameter into one diagnostic, for example with location as the first node to be removed, and `AdditionalLocations[0]` as the parameter to be replaced, and remaining `AdditionalLocations` as other nodes to be removed.

RikkiGibson replied:
> We need a way to only try to get this `parameterSyntax` if it wasn't already replaced.

I was thinking of just doing this with a `PooledHashSet<SyntaxNode>`.

> Or, consolidate diagnostics related to the same parameter into one diagnostic

This also seems doable, but maybe cumbersome on the analyzer side? Also, it means the code fix experience will be to "remove all null checks for the parameter"--versus, say, use the fixer to handle this one null check, then notice that the check later on in the code is erroneous and make an edit manually. Perhaps the parameter was reassigned and then null-tested, and we were throwing `ArgumentNullException` when we maybe should have been throwing a different kind of exception.

RikkiGibson replied:
Have resolved by adding a set of locations and avoiding replacing the node at the same location multiple times.

Youssef1313 commented on src/Analyzers/CSharp/CodeFixes/UseParameterNullChecking/CSharpUseParameterNullCheckingCodeFixProvider.cs:
```diff
@@ -47,6 +48,9 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
+            // Tracking parameters which have already been fixed by a fix-all operation.
+            // This avoids crashing the fixer when the same parameter is null-tested multiple times.
+            using var _ = PooledHashSet<Location>.GetInstance(out var fixedParameterLocations);
```
Do we need to `fixedParameterLocations.Free()` or similar at the end?

RikkiGibson replied:
I think `using var` handles freeing for us.

RikkiGibson commented:
The [fix all in solution results](https://github.com/RikkiGibson/roslyn/commit/9b0bb107b8180d2341a656764a13655cd6d5217b) for Roslyn is an interesting read. 😊

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,294 @@
+    {
+        private const string ArgumentNullExceptionName = $"{nameof(System)}.{nameof(ArgumentNullException)}";
+
+        public CSharpUseParameterNullCheckingDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseParameterNullCheckingId,
+                   EnforceOnBuildValues.UseParameterNullChecking,
+                   CSharpCodeStyleOptions.PreferParameterNullChecking,
+                   CSharpAnalyzersResources.Use_parameter_null_checking,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Null_check_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
```
@RikkiGibson For the `csharp_style_prefer_parameter_null_checking` not showing in the test, I think there is a base constructor overload that takes a language name (`LanguageNames.CSharp`) this could probably be the problem? I'm not sure.

Youssef1313 replied:
Actually I might be calling the constructor that accepts a language, but the language passed is `CSharpAnalyzersResources.Use_parameter_null_checking`.

Consider also instantiating a `LocalizableResourceString` for `Use_parameter_null_checking`

RikkiGibson replied:
I'm changing the base call using the `PreferNotPattern` analyzer as a starting point.

Thanks for the tip, it completely slipped under my radar 😄

Youssef1313 commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -23,7 +23,7 @@ public CSharpUseParameterNullCheckingDiagnosticAnalyzer()
             : base(IDEDiagnosticIds.UseParameterNullCheckingId,
                    EnforceOnBuildValues.UseParameterNullChecking,
                    CSharpCodeStyleOptions.PreferParameterNullChecking,
-                   CSharpAnalyzersResources.Use_parameter_null_checking,
```
Is Use_parameter_null_checking no longer used? I think it should be.

RikkiGibson replied:
This is odd. I'm going to just take some time later and ensure that this analyzer/fixer does something similar as UseNotPattern.

RikkiGibson replied:
`CSharpAnalyzersResources.Use_parameter_null_checking` is still used as the title of the code action. I went through and tried to imitate the `UseNotPattern` fixer and related analyzer as closely as possible. We should be good now.

CyrusNajmabadi commented:
The way to fix this issue is to preprocess the full list of diagnostics you are fixing. So you'll group them by parameter and then fix the parameter once and the N associated statements at the same time.

RikkiGibson commented:
It seems like the current approach of gathering and checking parameter locations in a set is adequate and doesn't require a separate initial pass over the diagnostics.

Youssef1313 reviewed (commented):
Worth adding a test like this?

```
using System;

class C<T> where T : struct
{
   public void M(T? s!!) { }
}
```

RikkiGibson commented:
See `TestOnStructConstrainedTypeParameter`

RikkiGibson commented:
@CyrusNajmabadi I believe I have addressed all your comments. Please let me know if you have any others.

CyrusNajmabadi commented:
Looking now :)

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,296 @@
+                    return;
+                }
+
+                var argumentNullException = compilation.GetTypeByMetadataName(ArgumentNullExceptionName);
```
```suggestion
                var argumentNullException = compilation.GetBestTypeByMetadataName(ArgumentNullExceptionName);
```

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,296 @@
+                if (parameter is not null
+                    && (!parameter.Type.IsValueType
+                        || parameter.Type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T
+                        || parameter.Type.TypeKind is TypeKind.Pointer or TypeKind.FunctionPointer)
```
extract into helper plz (can be local function or method).  mixing the && || and ! is too much for my brain :)

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,296 @@
+                    // if (param is null) { throw new ArgumentNullException(nameof(param)); }
+                    // if (object.ReferenceEquals(param, null)) { throw new ArgumentNullException(nameof(param)); }
+                    case IfStatementSyntax ifStatement:
+                        ExpressionSyntax left, right;
```
i would extract the entire if-handling code to it's own helper.

RikkiGibson replied:
I think I won't make this change due to lack of uniformity with the ExpressionStatement case. See https://github.com/dotnet/roslyn/pull/58182#discussion_r782508142

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,296 @@
+                        return parameterInBinary;
+
+                    // this.field = param ?? throw new ArgumentNullException(nameof(param));
+                    case ExpressionStatementSyntax
```
similarly, can you extract this case out to a helper?

RikkiGibson replied:
I found that since much of the complexity was in the pattern itself, the extracted logic didn't feel sufficiently complex to justify a single-use helper.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,296 @@
+                    return false;
+                }
+
+                return argumentNullExceptionStringConstructor.Equals(semanticModel.GetSymbolInfo(exceptionCreation, cancellationToken).Symbol);
```
consider reordering.  it's a personal nit, but it seems more sensible to first see if you're doing `new ArgumntNullException` *then* check the argument.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,1453 @@
+{
+    using VerifyCS = CSharpCodeFixVerifier<CSharpUseParameterNullCheckingDiagnosticAnalyzer, CSharpUseParameterNullCheckingCodeFixProvider>;
+
+    public class UseParameterNullCheckingTests
```
are there tests for records?

RikkiGibson replied:
Are you referring to record primary constructors? Currently I don't think there's a way to get this particular code fix to activate on primary constructors. The "add null check" refactoring should work for it, but it doesn't. Opened #58779

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,1453 @@
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestNotEqualitySwapped()
+        {
+            var testCode = @"using System;
```
possibly comment that this is just because we don't support this scenario yet.

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:
```diff
@@ -0,0 +1,1453 @@
+                    }
+                },
+                LanguageVersion = LanguageVersionExtensions.CSharpNext
+            }.RunAsync();
```
add test for:

```
public C(
      string x // comment
    , string y // comment,
    , string z // comment
    )
{
}
```

RikkiGibson replied:
Good test case! We actually put the `!!` on the next line here. It feels like the trailing trivia of the identifier should move over to the `!!`. Is there an existing helper that lets me "insert a token and move over another token's trivia" in this way?

CyrusNajmabadi replied:
nope.  you'd have to write the logic yourself.  note: i'm also ok with you leaving that out of this PR and waiting until tehre is user feedback :)

CyrusNajmabadi commented on src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:
```diff
@@ -166,6 +166,7 @@ internal static class IDEDiagnosticIds
         public const string SimplifyPropertyPatternDiagnosticId = "IDE0170";
 
         public const string UseTupleSwapDiagnosticId = "IDE0180";
+        public const string UseParameterNullCheckingId = "IDE0190";
```
```suggestion

        public const string UseParameterNullCheckingId = "IDE0190";
```
